lib/que/job_buffer.spec.rb

```diff
@@ -201,6 +201,50 @@ def new_metajob(key)
     end
   end
 
+  describe "push and shift" do
+    it "should not deadlock" do
+      job_buffer # Pre-initialize to avoid race conditions.
+
+      concurrency = 4
+      jobs_count_per_thread = 10_000
+
+      push_thread = Thread.new do
+        (concurrency * jobs_count_per_thread).times do
+          job_buffer.push(*job_array[0])
+        end
+      end
+
+      shift_threads = concurrency.times.map do
+        Thread.new do
+          jobs_count_per_thread.times do
+            job_buffer.shift
+          end
+        end
+      end
+
+      deadlock_detected = false
+      unlock_thread = Thread.new do
+        sleep 10
+
+        deadlock_detected = true
+        push_thread.kill
+        shift_threads.each { |t| t.kill }
+      end
+
+      push_thread.join
+      (concurrency * jobs_count_per_thread).times do
+        job_buffer.push(*job_array[0])
+      end
+
+      shift_threads.each { |t| t.join }
+      unlock_thread.kill
+
+      if deadlock_detected
+        raise "Deadlock"
+      end
+    end
+  end
```

Comment:
This test takes a few seconds to run. We tried to develop a more targeted test for this issue but everything we tried ended up reaching into the internal implementation to such an extent that the test became very brittle. Reducing the number of jobs per thread decreases the chance of a deadlock (on `master`) so we think this configuration is the best we can do for now.